[WIP] Job Config

sbliven opened the pull request:
## Description

Initial implementation of Job Config (#841). WIP.

sbliven committed:
Stub jobconfig.ts class

sbliven committed:
Improve jobconfig definition

- Initial JobConfig and JobAction interfaces.
- One example implementation: LogJobAction
- Basic tests

sbliven committed:
Start emailaction.ts [WIP]

This will replace email functionality in the job service.

This is still not working or tested, and commented code in the
job controller and service needs to be migrated to actions.

nitrosx commented on src/jobs/jobs.controller.ts:
```diff
@@ -242,97 +242,101 @@ export class JobsController {
-    // returns job instance configuration
-    return jc[0];
+    // // returns job instance configuration
+    // return jc[0];
```
Can you explain why this code is commented?

sbliven replied:
Just commenting stuff to try to make it compile.

nitrosx commented on src/jobs/jobs.controller.ts:
```diff
@@ -242,97 +242,101 @@ export class JobsController {
+    //     );
+    //   }
+    // } 
+    // return res;
```
Same question as above: can you explain why this code is commented?

sbliven replied:
As above, Just commenting stuff to try to make it compile.

nitrosx commented on src/config/actions/emailaction.ts:
I though that we were moving emails outside SciCat?

sbliven replied:
This could be a plugin in the future. For now core still has the nodemailer dependency so I figured it was a good prototype action implementation.

nitrosx commented on src/casl/casl-ability.factory.ts:
```diff
@@ -163,7 +163,7 @@ export class CaslAbilityFactory {
     //}
 
-    can(Action.Manage, Job);
+    can(Action.Manage, JobClass);
```
You do not need to change the casl authorization settings for jobs.
They have already been updated in the master branch according to the documentation:
https://github.com/SciCatProject/scicat-backend-next/blob/e3248e753fa5a01b1afb908a46db4f10e4e806e5/src/casl/casl-ability.factory.ts#L777

sbliven replied:
Ok, I guess this needs to be merged into release-jobs

nitrosx commented on src/casl/action.enum.ts:
I do not think that this is needed. 
CASL authorization have already been updated in master.

nitrosx reviewed (changes requested):
Please see my comments linked to the files

sbliven committed:
Merge remote-tracking branch 'origin/release-jobs' into issue-841

sbliven committed:
Minor changes from meeting

- Remove JobType enum. jobTypes are now strings to allow customization
- Rename `type` field to distinguish `JobConfig.jobType` from 
  `JobAction.actionType`

sbliven committed:
Implement job actions in the jobs controller

Also:
- Fix EmailJobAction. I'm not sure how nestjs MailerService works;
  just use nodemailer directly for now.
- Rename some remaining `type` fields
- TODO `configure()` is currently synchronous and contains promises.
  Maybe should be changed to be async or at least cached?

8 compilation errors remain. In particular, it's currently inconsistent whether `ActionJob.actionType` should be static.

sbliven committed:
Remove comments from previous commit

sbliven commented:
This branch is getting rather complicated. I think we need to
- Merge master into release-jobs to get the latest casl commits
- Rebase & squash issue-841 on the new release-jobs
- Close this PR and open a new on with the squashed version
- Merge quickly so that development on release-jobs can progress.

Maybe we need to do this as a zoom session.

sbliven committed:
Merge branch 'release-jobs' into issue-841

sbliven commented:
I'm closing this PR in favor of #1002, which is ready to be merged.
